datafusion/physical-plan/src/joins/dynamic_filters.rs

```diff
@@ -0,0 +1,267 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use arrow::array::AsArray;
+use arrow::compute::filter_record_batch;
+use arrow::datatypes::DataType;
+use arrow::record_batch::RecordBatch;
+use datafusion_common::{exec_err, DataFusionError};
+use datafusion_expr::Accumulator;
+use datafusion_expr::Operator;
+use datafusion_functions_aggregate::min_max::{MaxAccumulator, MinAccumulator};
+use datafusion_physical_expr::expressions::{BinaryExpr, Column, Literal};
+use datafusion_physical_expr::PhysicalExpr;
+use hashbrown::HashSet;
+use parking_lot::Mutex;
+use std::fmt;
+use std::sync::Arc;
+
+pub struct DynamicFilterInfo {
+    columns: Vec<Arc<Column>>,
+    build_side_names: Vec<String>,
+    inner: Mutex<DynamicFilterInfoInner>,
+}
+
+struct DynamicFilterInfoInner {
+    max_accumulators: Vec<MaxAccumulator>,
+    min_accumulators: Vec<MinAccumulator>,
```

Comment:
I wonder if we can just use `batches: Vec<Vec<Arc<dyn Array>>>` here and compute any filters from that?

Reply:
This makes it pretty easy to support additional  filters. It's pretty easy to add `InListExpr` filter based on this (I have some code locally).

Reply:
I agree, I'll refactor it, after that maybe I can combine your commit.

Reply:
Cool, after changes have cooled down a bit I can update and share my branch.
Adding something like that on top is actually quite easy after making the `make_set` helper public:
```
+                let col =
+                    Arc::<datafusion_physical_expr::expressions::Column>::clone(column);
+                let batches: Vec<&dyn Array> = inner.batches[i]
+                    .iter()
+                    .map(|x| x as &dyn Array)
+                    .collect();
+
+                // use inlist rather than min/max, use some threshold to avoid big inputs
+                // TODO: big inputs could be using bloom filter
+                if batches.iter().map(|x|x.len()).sum::<usize>() < 100 {
+                    let batches = concat(batches.as_slice())?;
+                    let set = make_set(&batches)?;
+    
+                    let unique_condition = InListExpr::new(col, vec![], false, Some(set));
+                    condition = Arc::new(BinaryExpr::new(
+                        Arc::new(unique_condition),
                         Operator::And,
-                        range_condition,
-                    ))
-                        as Arc<dyn PhysicalExpr>)),
-                    None => Ok(Some(range_condition)),
+                        condition,
+                    ));
                 }
+                Ok(Some(condition))
```

Reply:
Something that would be fun to try is creating a `PhysicalExpr` from the `JoinHashMap` - that way we can build a good filter without any overhead.

Reply:
I can do that

Reply:
I was trying to do this , but since hash function is one-way function, I can only combine the record batches to get original value... I didn't see the benefits here, I think one way to reduce overhead is just supporting max/min filter here(as what duckdb does)... If there is a better way, please tell me. I am getting a little confused here. Thanks

Reply:
So the idea is just to filter out values where the hash isn't present in the hashmap, that way we can filter out most values.

It's fine to focus on min/max first, the positive side is that evaluation is fast, but it won't filter out a lot of values if the min/max range from the left side is as big as the right side (regardless of number of values being much smaller).

Reply:
> So the idea is just to filter out values where the hash isn't present in the hashmap, that way we can filter out most values.
I got it, I was think about rebuild original array using hashmap and records 😅...... I'll add the logic

> It's fine to focus on min/max first, the positive side is that evaluation is fast, but it won't filter out a lot of values if the min/max range from the left side is as big as the right side (regardless of number of values being much smaller on left side).